### PR TITLE
Move tasks to next sprint on single click

### PR DESCRIPTION
**Pull request by duanxianbo**
image

### CODE REVIEW

File: src/client/sprint.client.ts [hunk 1]
```diff
@@ -107,6 +107,30 @@ export class SprintClient {
...
+        toSprintId: string,
+        taskIds: string[]
+    ): Promise<RemoteTask[]> {
+        return this.graphQLClient
+            .mutate({
+                mutation: gql`
+                    mutation moveTasksToSprint($fromSprintId: ID!, $toSprintId: ID!, $taskIds: [ID!]!) {
+                        moveTasksToSprint(fromSprintId: $fromSprintId, toSprintId: $toSprintId, taskIds: $taskIds) {
+                            id
+                            isPlanned
```
Comment (naturetime10) [inline src/client/sprint.client.ts]: I think this is not needed. `IsPlanned` shouldn't change when moving between sprints.

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 2]
```diff
@@ -53,39 +54,48 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+                            {this.state.sprints.length > 0 && (
+                                <div className={styles.Divider} />
+                            )}
+                            <div className={styles.Sprints}>
+                                {this.state.sprints.map(this.renderSprint)}
+                            </div>
+                        </div>
+                    )}
+                </div>
+
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx:64]:
```suggestion
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 3]
```diff
@@ -158,6 +168,60 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
         });
     };
 
+    private canMoveTasksToNextSprint = () => {
+        const currentSprint = this.graphSource.currentSprint();
+
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 4]
```diff
@@ -158,6 +168,60 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+    private getNextSprint(currentSprint: Sprint) {
+        const currentIndex = this.state.sprints.findIndex(
+            (sprint) => sprint.id === currentSprint.id,
+        );
+
+        if (currentIndex === 0) {
+            return null;
+        }
+
+        return this.state.sprints[currentIndex - 1];
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: I would recommend we find the next sprint by first make a copy of sprints array with `map`. Then find the index of the current sprint. We can't assume `sprints` are sorted.

File: src/data/task.filter.ts [hunk 5]
```diff
@@ -20,6 +20,8 @@ export const inProgress: Filter<Task> = (task: Task): boolean =>
     task.status === 'IN_PROGRESS';
 export const delivered: Filter<Task> = (task: Task): boolean =>
     task.status === 'DELIVERED';
+export const nonDelivered: Filter<Task> = (task: Task): boolean =>
+    task.status !== 'DELIVERED';
```
Comment (naturetime10) [inline src/data/task.filter.ts]: Can we rename `nonDelivered` to unfinished, which I feel is more natural.

File: src/storage/syncer/stateSyncer.ts [hunk 6]
```diff
@@ -669,6 +680,44 @@ export class StateSyncer {
...
+    ): Promise<AppState> {
+        const remoteTasks = await this.sprintClient.moveTasksToSprint(
+            `${fromSprintId}`,
+            `${toSprintId}`,
+            taskIds.map(String),
+        );
+
+        return remoteTasks.reduce((appState, remoteTask) => {
+            const taskId = toInt(remoteTask.id);
+
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]:
```suggestion
```

File: src/storage/syncer/stateSyncer.ts [hunk 7]
```diff
@@ -669,6 +680,44 @@ export class StateSyncer {
         return appState;
     }
 
+    private async _moveTasksToSprint(
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]: Are we using this function somewhere else? If not, why are we creating a helper function?

File: src/storage/syncer/stateSyncer.ts [hunk 8]
```diff
@@ -669,6 +680,44 @@ export class StateSyncer {
...
+                    ),
+            );
+
+            const relation: SprintTaskRelationState = {
+                sprintId: toSprintId,
+                taskId,
+            };
+            appState.sprintTaskRelations =
+                appState.sprintTaskRelations.concat(relation);
+
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]:
```suggestion
```

File: src/storage/syncer/stateSyncer.ts [hunk 9]
```diff
@@ -669,6 +680,44 @@ export class StateSyncer {
...
+        taskIds: number[],
+    ): Promise<AppState> {
+        const remoteTasks = await this.sprintClient.moveTasksToSprint(
+            `${fromSprintId}`,
+            `${toSprintId}`,
+            taskIds.map(String),
+        );
+
+        return remoteTasks.reduce((appState, remoteTask) => {
+            const taskId = toInt(remoteTask.id);
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]: Maybe extra to helper function and call it `updateTaskSprint`?

**Comment by naturetime10** [comment]
Thanks for creating this PR! It's challenging to understand the final effect of HTML & CSS changes in this PR. Can you add a screenshot in PR description?

File: src/storage/syncer/stateSyncer.ts (hunk 9 above)
Comment (naturetime10) [reply src/storage/syncer/stateSyncer.ts]: This is a good use of reduce function. Moving forwards, we can refactor our codebase to use this pattern.

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/components/internal/SprintsHeaderSection.component.tsx (hunk 4 above)
Comment (duanxianbo) [reply src/components/internal/SprintsHeaderSection.component.tsx]: done

File: src/storage/syncer/stateSyncer.ts (hunk 7 above)
Comment (duanxianbo) [reply src/storage/syncer/stateSyncer.ts]: moved

File: src/storage/syncer/stateSyncer.ts (hunk 9 above)
Comment (duanxianbo) [reply src/storage/syncer/stateSyncer.ts]: done

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 10]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
         });
     };
 
+    private canMoveTasksToNextSprint = () => {
+        const currentSprint = this.graphSource.currentSprint();
+
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 11]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+        if (!currentSprint) {
+            return false;
+        }
+
+        const nonDelivereTasks = (currentSprint.tasks || []).filter(
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
        const unfinishedTasks = (currentSprint.tasks || []).filter(
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 12]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+            unfinished,
+        );
+
+        if (!nonDelivereTasks.length) {
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
        if (!unfinishedTasks.length) {
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 13]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+        if (!nextSprint) {
+            return false;
+        }
+
+        return true;
+    };
+
+    private onMoveTasksToNextSprint = () => {
+        if (!this.canMoveTasksToNextSprint()) {
+            alert('can not move to next sprint');
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
            alert('can not move tasks to next sprint');
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 14]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+            return;
+        }
+
+        const currentSprint = this.graphSource.currentSprint()!;
+        const nextSprint = this.getNextSprint(currentSprint)!;
+        const nonDelivereTasks = (currentSprint.tasks || []).filter(
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
        const unfinishedTasks = (currentSprint.tasks || []).filter(
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 15]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+            unfinished,
+        );
+        const nonDelivereTaskIds = nonDelivereTasks.map((task) => task.id);
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
        const unfinishedTaskIds = unfinishedTasks.map((task) => task.id);
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 16]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+        this.stateSyncer.moveTasksToSprint(
+            currentSprint.id,
+            nextSprint.id,
+            nonDelivereTaskIds,
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
            unfinishedTaskIds,
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 17]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+        );
+    };
+
+    private getNextSprint(currentSprint: Sprint) {
+        let sprints = [...this.state.sprints]
+        sprints = orderByStartTimeDesc(sprints)
+
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
```

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 18]
```diff
@@ -158,6 +175,63 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+        const currentIndex = sprints.findIndex(
+            (sprint) => sprint.id === currentSprint.id,
+        );
+
+        if (currentIndex === 0) {
+            return null;
+        }
+
+        return sprints[currentIndex - 1];
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx:230]: Next PR: handle the case when next sprint doesn't exist

File: src/components/internal/shared.scss [hunk 19]
```diff
@@ -31,3 +31,26 @@ $sectionSpace: 10px;
         background-color: $primaryColor;
     }
 }
+
```
Comment (naturetime10) [inline src/components/internal/shared.scss:34]:
```suggestion
```

File: src/storage/syncer/stateSyncer.ts [hunk 20]
```diff
@@ -652,6 +651,60 @@ export class StateSyncer {
...
+    }
+
+    private _updateTaskSprint(
+        appState: AppState,
+        remoteTask: RemoteTask,
+        fromSprintId: number,
+        toSprintId: number,
+    ): AppState {
+        const taskId = toInt(remoteTask.id);
+
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]:
```suggestion
```

File: src/storage/syncer/stateSyncer.ts [hunk 21]
```diff
@@ -652,6 +651,60 @@ export class StateSyncer {
...
+        if (taskId === undefined) {
+            return appState;
+        }
+        appState = mergeTask(appState, remoteTask);
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]:
```suggestion
        }
        
        appState = mergeTask(appState, remoteTask);
```

File: src/storage/syncer/stateSyncer.ts [hunk 22]
```diff
@@ -652,6 +651,60 @@ export class StateSyncer {
...
+                ),
+        );
+
+        const relation: SprintTaskRelationState = {
+            sprintId: toSprintId,
+            taskId,
+        };
+        appState.sprintTaskRelations =
+            appState.sprintTaskRelations.concat(relation);
+
```
Comment (naturetime10) [inline src/storage/syncer/stateSyncer.ts]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**